src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs

```diff
@@ -178,5 +153,97 @@ public bool ManualSettingsOnly
 
             return useProxy;
         }
+
+        private void DetectScriptLocation()
+        {
+            if (_scriptDetectionFailed || AutoConfigUrl != null)
+            {
+                return;
+            }
+
+            WinHttpTraceHelper.Trace("WinInetProxyHelper.DetectScriptLocation: Attempting discovery PROXY_AUTO_DETECT_TYPE_DHCP.");
+            AutoConfigUrl = GetAutoProxyConfigUrl(Interop.WinHttp.AutoDetectType.Dhcp);
+
+            if (AutoConfigUrl == null)
+            {
+                WinHttpTraceHelper.Trace("WinInetProxyHelper.DetectScriptLocation: Attempting discovery AUTO_DETECT_TYPE_DNS_A.");
+                AutoConfigUrl = GetAutoProxyConfigUrl(Interop.WinHttp.AutoDetectType.DnsA);
+            }
+            else
+            {
+                // Auto-detect succeeded.
+                return;
+            }
+
+            if (AutoConfigUrl == null)
+            {
+                WinHttpTraceHelper.Trace("WinInetProxyHelper.DetectScriptLocation: Auto discovery failed.");
+                AutoDetect = false;
+                _scriptDetectionFailed = true;
```

Comment:
Why do we need _scriptDetectionFailed at all?  It seems like it's true iff AutoConfigUrl.

Reply:
And do we need AutoDetect? In what situation will that be true but not have a url, and what are we supposed to do at that point?

Reply:
> And do we need AutoDetect? In what situation will that be true but not have a url, and what are we supposed to do at that point?

It's common that `AutoDetect` is true while `AutoConfigUrl` is null. This can be caused by, specifying `Automatically detect settings` while not providing a location for `automatic configuration script` in the IE configuration.

If `AutoConfigUrl` is null (no configuration script is specified), and `AutoDetect` is true, the next step is that we will use WPAD to find the script location. (`WinHttpGetProxyForUrl`/`WinHttpDetectAutoProxyConfigUrl`).

Since within current function, we are calling into `WinHttpDetectAutoProxyConfigUrl` to help us find the script. If it fails, we will update the `AutoDetect` to false, so that we will not calling into `WinHttpGetProxyForUrl` anymore. (The purpose of this PR, fix the perf hit for `WinHttpGetProxyForUrl`).

(we will call into `WinHttpGetProxyForUrl` if `if (_proxyHelper != null && _proxyHelper.AutoSettingsUsed)` -> `AutoSettingsUsed: return AutoDetect || !string.IsNullOrEmpty(AutoConfigUrl);`)